### DISCUSSION TITLE
Required validation not working using annotations

### DISCUSSION THREAD

**Discussion by SeriousKen** [livewire/livewire#5782] [7. Bugs (archived)]
Just trying out version 3 and came across an issue. I created the following component:

``` php
<?php

namespace App\Livewire;

use Livewire\Attributes\Rule;
use Livewire\Component;

class DemoForm extends Component
{
    #[Rule('required')]
    public $firstname = '';

    #[Rule('required')]
    public $lastname = '';

    public function render()
    {
        return view('livewire.demo-form');
    }

    public function save()
    {
        // Save logic goes here
    }
}
```

with the following template:

``` blade
<div>
    <form wire:submit='save'>
        <div>
            <label>First name</label>
            <input wire:model='firstname' />

            @error('firstname')
                <span>Error: {{ $message }}</span>
            @enderror
        </div>
        <div>
            <label>Last name</label>
            <input wire:model='lastname' />

            @error('lastname')
                <span>Error: {{ $message }}</span>
            @enderror
        </div>
        <div>
            <button type="submit">Submit</button>
        </div>
    </form>

    <p>{{ $firstname }} {{ $lastname }}</p>
</div>
```

Submitting the form blank doesn't display any errors, however, if I put `$this->validate()` into the save method and add `onUpdate: false` to the annotations then it works as expected.

**Comment by joshhanley** [comment]
@SeriousKen validation will only run on per property if your using .live or something like that. Otherwise you have to call `$this->validate()`.

Hope this helps!

**Reply by SeriousKen** [reply]
I'm just following the documentation [here](https://livewire.laravel.com/docs/forms) under the heading _Adding validation_ which makes no mention of requiring the .live modifier.

**Reply by joshhanley** [reply]
@SeriousKen ah ok, thanks. Docs probably need updating. Trying adding `$this->validate();` to the save method.

**Reply by SeriousKen** [reply]
The documentation [here](https://livewire.laravel.com/docs/validation#rule-attributes) also seems to be pretty clear that the intended behaviour is for validation to run without the need to call `$this->validate` for `#[Rule]` attributes. It seems to work if the validation is anything but `required`. If I replace `required` with `min:3` for example then when submitting the form I get form errors back, but it does seem like the `save` method is still called even if there are form errors.

**Reply by joshhanley** [reply]
@SeriousKen ah right, I didn't realise that, thanks for letting me know 🙂 hmm I wonder why it's not working with required... Any chance you could submit a PR with a failing test? Thanks!

**Reply by SeriousKen** [reply]
I've tried to do this, but as I'm using WSL2 on Windows 11 I'm having trouble getting the testing set up. There seems to be an issue with Chrome and WSL that's getting in the way of running Dusk tests and I'm not that familiar with Dusk.

**Comment by nam-co** [comment]
Upon testing, it seems it only works when both .live and $this->validate() are present, @joshhanley will protected $rules = [] always be supported? they seem more compact an easier

**Reply by joshhanley** [reply]
@nam-co hmm interesting, any chance you could submit a failing test PR? Yeah I don't think `$rules` are going away as there are still thinks that attribute rules can't do (like exclude an instantiated ID from an exists check).

**Reply by nam-co** [reply]
Hi @joshhanley if I was half as good as you are I would definitely do it 😀 , but I don’t know how to do it, but I did find in beta.3, you definitely need the $this->validate() , after is inserted, wire:model alone requires another trigger to show the error, .live is instant and .blur the click outside, hope this helps

**Reply by joshhanley** [reply]
@nam-co I believe in you, I reckon you could manage it! I'm happy to support you if you run into issues. 🙂 Have a look at our old V2 contribution guide for details https://laravel-livewire.com/docs/2.x/contribution-guide

We're just flat out at the moment with the amount of issues, so the quickest way to get stuff fixed is with a failing test, as we don't have time ourselves to write tests for each issue.

**Reply by nam-co** [reply]
I will try to do it tonight 👍

**Reply by joshhanley** [reply]
@nam-co thanks, that's much appreciated! 🙂

**Comment by marksalmon** [comment]
Hi @joshhanley, I'm trying to get the repo running locally so that I can submit a failing test but I'm having a problem with ```npm install```:

<img width="1231" alt="image" src="https://github.com/livewire/livewire/assets/7000886/4739ee3e-979f-4180-aa12-bcfa81ac01e3">


I have got past that by removing that line from package.json, but now I cannot run the dusk tests. 

I have followed all of the steps in the (v2) contribution guide but the dusk tests are failing due to ```ERR_CONNECTION_REFUSED``` and the chrome window is showing ```127.0.0.1 refused to connect```

I'm afraid I'm not that familiar with dusk, but I guess I'm not runnnig a web server where dusk is expecting one (I am using Herd on MacOS)

Can you point me in the right direction?

**Reply by nuernbergerA** [reply]
@marksalmon see https://github.com/livewire/livewire/discussions/5837#discussioncomment-6515864

**Reply by marksalmon** [reply]
Thanks, but if I follow that to fix the history package, I get the error for the navigate package, and if i fix the navigate package I get the history package error again. Is is possible to have multiple npm links?

**Reply by nuernbergerA** [reply]
@marksalmon `npm link A B C`

**Reply by marksalmon** [reply]
Thanks very much. That has solved the npm problem. Now to get the dusk tests working...

**Reply by PhiloNL** [reply]
> Thanks very much. That has solved the npm problem. Now to get the dusk tests working...
> 
> 

This is related to Herd / testbench. This should be fixed in a new release https://github.com/beyondcode/herd-community/issues/21

**Reply by marksalmon** [reply]
Excellent. I'm edging closer.

Now I can get the dusk tests to run, but the browser just hangs and the test timeout:

<img width="836" alt="image" src="https://github.com/livewire/livewire/assets/7000886/d794ff81-a082-4036-aa51-4f0231ef54e1">

<img width="978" alt="image" src="https://github.com/livewire/livewire/assets/7000886/af48bec7-3a6d-4387-b8d8-d6d232e0ce6d">
